File: web/src/api/ssl.ts
import { apiFetch } from "./client";

export interface SSLStatus {
  active: boolean;
  mode: string;
  domain: string;
  subject: string | null;
  sans: string[] | null;
  not_before: string | null;
  not_after: string | null;
  days_until_expiry: number | null;
}

export interface SSLUploadResult {
  status: "restarting";
  domain: string;
}

export interface SSLRemoveResult {
  status: "restarting";
}

export function fetchSSLStatus(): Promise<SSLStatus> {
  return apiFetch<SSLStatus>("/api/v1/system/ssl/status");
}

export function uploadSSLPfx(
  file: File,
  password: string,
  domain: string
): Promise<SSLUploadResult> {
  const fd = new FormData();
  fd.append("method", "pfx");
  fd.append("domain", domain);
  fd.append("file", file);
  fd.append("pfx_password", password);
  return apiFetch<SSLUploadResult>("/api/v1/system/ssl/upload", { method: "POST", body: fd });
}

export function uploadSSLPemFile(certFile: File, domain: string): Promise<SSLUploadResult> {
  const fd = new FormData();
  fd.append("method", "pem_file");
  fd.append("domain", domain);
  fd.append("file", certFile);
  return apiFetch<SSLUploadResult>("/api/v1/system/ssl/upload", { method: "POST", body: fd });
}

export function uploadSSLPemPaste(pemText: string, domain: string): Promise<SSLUploadResult> {
  const fd = new FormData();
  fd.append("method", "pem_paste");
  fd.append("domain", domain);
  fd.append("pem_text", pemText);
  return apiFetch<SSLUploadResult>("/api/v1/system/ssl/upload", { method: "POST", body: fd });
}

export function removeSSL(): Promise<SSLRemoveResult> {
  return apiFetch<SSLRemoveResult>("/api/v1/system/ssl", { method: "DELETE" });
}

export async function pollSSLHealthy(
  timeoutMs = 30_000,
  intervalMs = 1_000
): Promise<boolean> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    try {
      const url = `https://${window.location.hostname}:8443/health`;
      const res = await fetch(url, { mode: "no-cors" });
      if (res.ok || res.type === "opaque") return true;
    } catch {
      /* not yet up */
    }
    await new Promise((r) => setTimeout(r, intervalMs));
  }
  return false;
}

File: api/routers/ssl.py
import asyncio
import os
import signal
from typing import Optional

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile
from fastapi.responses import JSONResponse

from api.dependencies import require_admin, AuthUser
from services import ssl_service
from services.ssl_service import SSLValidationError

router = APIRouter()


def _require_container_mode() -> None:
    if os.environ.get("ZS_CONTAINER_MODE") != "1":
        raise HTTPException(status_code=503, detail="ssl_container_only")


async def _delayed_restart() -> None:
    await asyncio.sleep(2)
    os.kill(os.getpid(), signal.SIGTERM)


@router.post("/api/v1/system/ssl/upload", tags=["System"])
async def upload_ssl(
    method: str = Form(...),
    domain: str = Form(...),
    file: Optional[UploadFile] = File(default=None),
    pfx_password: str = Form(default=""),
    pem_text: str = Form(default=""),
    _: AuthUser = Depends(require_admin),
) -> dict:
    _require_container_mode()

    if method not in ("pfx", "pem_file", "pem_paste"):
        raise HTTPException(status_code=422, detail="method must be pfx, pem_file, or pem_paste")
    if method in ("pfx", "pem_file") and file is None:
        raise HTTPException(status_code=422, detail="file is required for pfx and pem_file methods")
    if method == "pem_paste" and not pem_text.strip():
        raise HTTPException(status_code=422, detail="pem_text is required for pem_paste method")

    try:
        if method == "pfx":
            bundle = ssl_service.process_pfx(await file.read(), pfx_password, domain)
        elif method == "pem_file":
            bundle = ssl_service.process_pem_bytes(await file.read(), domain)
        else:
            bundle = ssl_service.process_pem_text(pem_text, domain)
        ssl_service.save_bundle(bundle, domain)
    except SSLValidationError as e:
        return JSONResponse(status_code=400, content={"detail": e.code, "message": str(e)})

    asyncio.create_task(_delayed_restart())
    return {"status": "restarting", "domain": domain}


@router.get("/api/v1/system/ssl/status", tags=["System"])
def ssl_status(_: AuthUser = Depends(require_admin)) -> dict:
    s = ssl_service.get_status()
    return {
        "active": s.active,
        "mode": s.mode,
        "domain": s.domain,
        "subject": s.subject,
        "sans": s.sans,
        "not_before": s.not_before,
        "not_after": s.not_after,
        "days_until_expiry": s.days_until_expiry,
    }


@router.delete("/api/v1/system/ssl", tags=["System"])
async def remove_ssl(_: AuthUser = Depends(require_admin)) -> dict:
    _require_container_mode()
    ssl_service.remove_ssl()
    asyncio.create_task(_delayed_restart())
    return {"status": "restarting"}
